File: packages/contracts/src/timeline.ts
import type { JsonObject, Rational, Uuid } from "./common.js";

export type TimelineTrackKind = "video" | "audio" | "text" | "overlay";
export type BlendMode = "normal" | "multiply" | "screen" | "overlay";

export interface TimelineSource {
  readonly media_id: Uuid;
  readonly uri: string;
  readonly source_in: number;
  readonly source_duration: number;
}

export interface TimelineEffect {
  readonly type: string;
  readonly version: string;
  readonly parameters: JsonObject;
}

export interface TimelineClip {
  readonly id: Uuid;
  readonly track_id: Uuid;
  readonly start: number;
  readonly duration: number;
  readonly source: TimelineSource | null;
  readonly text: string | null;
  readonly speed: Rational;
  readonly opacity: number;
  readonly volume: number;
  readonly blend_mode: BlendMode;
  readonly effects: readonly TimelineEffect[];
  readonly supporting_segment_ids: readonly Uuid[];
  readonly supporting_claim_ids: readonly Uuid[];
}

export interface TimelineTrack {
  readonly id: Uuid;
  readonly kind: TimelineTrackKind;
  readonly name: string;
  readonly order: number;
  readonly exclusive: boolean;
  readonly muted: boolean;
  readonly clips: readonly TimelineClip[];
}

export interface TimelineTransition {
  readonly id: Uuid;
  readonly from_clip_id: Uuid;
  readonly to_clip_id: Uuid;
  readonly type: "cut" | "crossfade" | "dip_to_black" | "audio_crossfade";
  readonly duration: number;
}

export interface TimelineMarker {
  readonly id: Uuid;
  readonly position: number;
  readonly kind: "beat" | "proof" | "chapter" | "warning";
  readonly label: string;
  readonly reference_id: Uuid | null;
}

export interface SafeArea {
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
  readonly left: number;
}

export interface TimelineProject {
  readonly schema_version: "1.0";
  readonly id: Uuid;
  readonly project_id: Uuid;
  readonly width: number;
  readonly height: number;
  readonly fps: Rational;
  readonly timebase: Rational;
  readonly duration: number;
  readonly safe_area: SafeArea;
  readonly tracks: readonly TimelineTrack[];
  readonly transitions: readonly TimelineTransition[];
  readonly markers: readonly TimelineMarker[];
}

File: legacy/packages/contracts/python/src/gta_ai_contracts/timeline.py
from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import Field, model_validator

from .common import JsonValue, NonNegativeInt, PositiveInt, Rational, StudioModel, UnitScore


class TimelineSource(StudioModel):
    media_id: UUID
    uri: str
    source_in: NonNegativeInt
    source_duration: PositiveInt


class TimelineEffect(StudioModel):
    type: str
    version: str
    parameters: dict[str, JsonValue] = Field(default_factory=dict)


class TimelineClip(StudioModel):
    id: UUID
    track_id: UUID
    start: NonNegativeInt
    duration: PositiveInt
    source: TimelineSource | None = None
    text: str | None = None
    speed: Rational = Rational(numerator=1, denominator=1)
    opacity: UnitScore = 1
    volume: float = Field(default=1, ge=0, le=4)
    blend_mode: Literal["normal", "multiply", "screen", "overlay"] = "normal"
    effects: list[TimelineEffect] = Field(default_factory=list)
    supporting_segment_ids: list[UUID] = Field(default_factory=list)
    supporting_claim_ids: list[UUID] = Field(default_factory=list)


class TimelineTrack(StudioModel):
    id: UUID
    kind: Literal["video", "audio", "text", "overlay"]
    name: str
    order: NonNegativeInt
    exclusive: bool = True
    muted: bool = False
    clips: list[TimelineClip] = Field(default_factory=list)


class TimelineTransition(StudioModel):
    id: UUID
    from_clip_id: UUID
    to_clip_id: UUID
    type: Literal["cut", "crossfade", "dip_to_black", "audio_crossfade"]
    duration: NonNegativeInt


class TimelineMarker(StudioModel):
    id: UUID
    position: NonNegativeInt
    kind: Literal["beat", "proof", "chapter", "warning"]
    label: str
    reference_id: UUID | None = None


class SafeArea(StudioModel):
    top: UnitScore
    right: UnitScore
    bottom: UnitScore
    left: UnitScore

    @model_validator(mode="after")
    def validate_visible_area(self) -> "SafeArea":
        if self.top + self.bottom >= 1 or self.left + self.right >= 1:
            raise ValueError("safe area must leave a visible rectangle")
        return self


class TimelineProject(StudioModel):
    schema_version: Literal["1.0"] = "1.0"
    id: UUID
    project_id: UUID
    width: PositiveInt
    height: PositiveInt
    fps: Rational
    timebase: Rational
    duration: PositiveInt
    safe_area: SafeArea
    tracks: list[TimelineTrack]
    transitions: list[TimelineTransition] = Field(default_factory=list)
    markers: list[TimelineMarker] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_references_and_ranges(self) -> "TimelineProject":
        track_ids: set[UUID] = set()
        clip_ids: set[UUID] = set()
        for track in self.tracks:
            if track.id in track_ids:
                raise ValueError("timeline track ids must be unique")
            track_ids.add(track.id)
            ordered = sorted(track.clips, key=lambda clip: clip.start)
            for index, clip in enumerate(ordered):
                if clip.id in clip_ids:
                    raise ValueError("timeline clip ids must be unique")
                clip_ids.add(clip.id)
                if clip.track_id != track.id:
                    raise ValueError("clip track_id must match containing track")
                if clip.start + clip.duration > self.duration:
                    raise ValueError("clip exceeds timeline duration")
                if track.kind == "text" and not (clip.text and clip.text.strip()):
                    raise ValueError("text tracks require non-empty text")
                if track.exclusive and index > 0:
                    previous = ordered[index - 1]
                    if clip.start < previous.start + previous.duration:
                        pair_has_transition = any(
                            {transition.from_clip_id, transition.to_clip_id} == {previous.id, clip.id}
                            for transition in self.transitions
                        )
                        if not pair_has_transition:
                            raise ValueError("exclusive track clips overlap without transition")
        if any(transition.from_clip_id not in clip_ids or transition.to_clip_id not in clip_ids for transition in self.transitions):
            raise ValueError("transition references unknown clip")
        return self

